### DISCUSSION TITLE
Howto query REST APIs with paginated result sets?

### DISCUSSION THREAD

**Discussion by reneoelke** [duckdb/duckdb#14385] [Q&A]
Most RESTfull APIs have paginated result sets which use offset attributes to get all results in multiple requests. For example the HubSpot Contacts API (see https://legacydocs.hubspot.com/docs/methods/contacts/get_contacts).

Currently it is not possible to create 1 database table `hubspot_contacts` with all contacts. A maximum of 100 contacts from the first page of the result set can be used in this table (`count=100`).

Would it be possible to use the result attributes `has-more` and `vid-offset` to create a table with all contacts by multiple requests (until `has-more: false`)?

**Comment by Znote** [comment]
The link you provided has code samples in Python and Node JS on how to keep loading data from the  API endpoint. Just pass it along to DuckDB once all the pages and rows has been loaded?

Extended the sample code from Python, look at the bottom below:
`# Loading the full contact_list into a DuckDB Table:`

```py
# This example displays how to get all contacts from a HubID and paginate through them using the 'offset' parameter.
# The end result is a python list containing all parsed contacts. 

import requests
import json
import urllib

max_results = 500 
count = 100 
contact_list = []
get_all_contacts_url = "https://api.hubapi.com/contacts/v1/lists/all/contacts/all"
parameter_dict = {'count': count}
headers = {}

# Paginate your request using offset
has_more = True
while has_more:
	parameters = urllib.urlencode(parameter_dict)
	get_url = get_all_contacts_url + parameters
	r = requests.get(url= get_url, headers = headers)
	response_dict = json.loads(r.text)
	has_more = response_dict['has-more']
	contact_list.extend(response_dict['contacts'])
	parameter_dict['vidOffset']= response_dict['vid-offset']
	if len(contact_list) >= max_results: # Exit pagination, based on whatever value you've set your max results variable to. 
		print('maximum number of results exceeded')
		break
print('loop finished')

list_length = len(contact_list) 

print("You've succesfully parsed through {} contact records and added them to a list".format(list_length))

# Loading the full contact_list into a DuckDB Table:
import duckdb
from io import StringIO

table_contacts = duckdb.read_json(StringIO(json.dumps(contact_list)))
db_conn = duckdb.connect("database.duckdb")
db_conn.execute('CREATE TABLE hubspot_contacts AS FROM "table_contacts"')
db_conn.show()
db_conn.close()
# Explore and query table hubspot_contacts in database.duckdb as you wish
```

Note that the provided sample code has a max_results limit at 500, just increase that, or remove that check in the while loop to keep going indefinetely.

**Reply by reneoelke** [reply]
@Znote Thanks for you message. This is a good workaround for solving the issue with code.

I was thinking about a solution which could be solved with attributes for the `read_json` function or the `httpfs` extension. So that I can declare the `CREATE TABLE` statement once and then the rest will be done by the function or extension automatically.

Here is a pseudo code example:
```
CREATE TABLE customers AS 
SELECT
  *
FROM
  read_json(
    'https://api.hubapi.com/companies/v2/companies',
    auto_detect = true,
    httpfs_options = {
      pagination = true,
      pagination_parameter = "vidOffset",
      pagination_field = "vid-offset",
      pagination_condition = {
        has-more: true
      }
  )
;
```

**Reply by Znote** [reply]
Or perhaps a community driven httpapi extension that uses httpfs and json to support oauth tokens and typical nextpage request handlers. Not sure if that should be baked into a core extension.

**Reply by Znote** [reply]
Im still very new with duckdb (started playing with it a couple days ago). 

```sql
D CREATE OR REPLACE MACRO union_table(url, url_next) AS TABLE
  SELECT CONCAT(url, url_next+1) as next_url_column, * FROM read_json(CONCAT(url, url_next))
  UNION ALL 
  SELECT CONCAT(url, url_next+2) as next_url_column, * FROM read_json(CONCAT(url, url_next+1));
D 
D FROM union_table('https://jsonplaceholder.typicode.com/todos/', 1);
┌──────────────────────────────────────────────┬────────┬───────┬────────────────────────────────────┬───────────┐
│               next_url_column                │ userId │  id   │               title                │ completed │
│                   varchar                    │ int64  │ int64 │              varchar               │  boolean  │
├──────────────────────────────────────────────┼────────┼───────┼────────────────────────────────────┼───────────┤
│ https://jsonplaceholder.typicode.com/todos/2 │      1 │     1 │ delectus aut autem                 │ false     │
│ https://jsonplaceholder.typicode.com/todos/3 │      1 │     2 │ quis ut nam facilis et officia qui │ false     │
└──────────────────────────────────────────────┴────────┴───────┴────────────────────────────────────┴───────────┘
```

There might be a way to create a macro that loads json from url, turns it into a table, queries the data to generate url for the next page and loads that through union all until everything is loaded?

some recursive conditional union'ing :P

I miss ms-SQL/tsql stored procedures where we could create a query row cursor to generate dynamic sql code to execute.
